skfp/model_selection/splitters/pubchem_split.py

```diff
@@ -0,0 +1,447 @@
+    """
+    if n_jobs > 5:
+        n_jobs = 5
+
+    data = ensure_smiles(data)
+
+    if verbose > 0:
+        print("Converting SMILES to CIDs")
+    cids = run_in_parallel(_get_cid_for_smiles, data, n_jobs)
+
+    if verbose > 0:
+        print("Searching for CIDs literature")
+    years = run_in_parallel(_get_earliest_publication_date, cids, n_jobs)
+
+    return years
+
+
+def _get_cid_for_smiles(smiles: str) -> Optional[str]:
+    """
+
+    Parameters:
+    ------------
+    smiles: str
+        SMILES string
+
+    Return
+    --------
+        PubChem compound identifier if  else None
+
+    """
+    url = f"https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/smiles/{quote(smiles)}/cids/JSON"
+
+    response = requests.get(url, timeout=10)
+    response_status = response.status_code
+
+    if response_status != 200:
+        warnings.warn(
+            f"PUG error \n, {url} \n status code {response_status}, \n check your input"
+        )
+
+    response_json = response.json()
+
+    try:
+        cid = response_json["IdentifierList"]["CID"][0]
+        cid = str(cid) if cid else None
+    except (KeyError, ValueError):
+        cid = None
+    return cid
+
+
+def _get_earliest_publication_date(
+    cid: Optional[str], n_trials: int = 5
+) -> Optional[int]:
+    """
+    Get the date of the earliest publication from PubChem where a given molecule appears.
+    There are molecules without publications, for which we return None.
+
+    """
+    if not cid:
+        return None
+
+    query = {
+        "download": "*",
+        "collection": "literature",
+        "where": {"ands": [{"cid": str(cid)}]},
+        "order": ["articlepubdate,asc"],
+        "start": 0,
+        "limit": 3,  # fetch a few, just in case
+        "downloadfilename": f"pubchem_cid_{cid}_literature",
+        "nullatbottom": 1,
+    }
+    base_url = "https://pubchem.ncbi.nlm.nih.gov/sdq/sdqagent.cgi"
+    params = {"infmt": "json", "outfmt": "json", "query": json.dumps(query)}
+
+    def get_publication_date(response: list[dict]) -> Optional[int]:
+        article_pub_date = None
+        for entry in response:
+            if "articlepubdate" in entry:
+                return int(entry["articlepubdate"][:4])  # get year of publication
+        return article_pub_date
+
+    trial = 0
+    while trial < n_trials:
+        try:
+            response = requests.get(base_url, params=params, timeout=10).json()
+            if isinstance(response, list):
+                if len(response) >= 1:
+                    date = get_publication_date(response)
+                    return date
+                else:
+                    # no literature found
+                    return None
+            else:
+                # most probably "Server too busy" error
+                time.sleep(1)
+        except requests.ConnectionError:
+            # most probably unstable PubChem network
+            time.sleep(1)
+        trial += 1
+
+    return None
```

Comment:
If all trials failed, I think we should raise an error, because it's not `None`, but rather a failure because of PubChem servers